// @fastly/next-compute-js/build/src/core/util.js
"use strict";
/*
 * Copyright Fastly, Inc.
 * Licensed under the MIT license. See LICENSE file for details.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.doShutdown = exports.onShutdown = exports.doInit = exports.onInit = exports.isTest = exports.removeAllActions = exports.removeAction = exports.doAction = exports.addAction = void 0;
class ActionsManager {
    constructor() {
        this._actionsMap = new Map();
        this._dirtyKeys = new Set();
    }
    addAction(name, priority, fn) {
        let actions = this._actionsMap.get(name);
        if (actions == null) {
            actions = [];
            this._actionsMap.set(name, actions);
        }
        actions.push({ priority, fn });
        this._dirtyKeys.add(name);
    }
    sort() {
        for (const key of [...this._dirtyKeys]) {
            const actions = this._actionsMap.get(key);
            this._dirtyKeys.delete(key);
            if (actions == null) {
                continue;
            }
            actions.sort((a, b) => {
                return a.priority - b.priority;
            });
        }
    }
    doAction(name, ...args) {
        this.sort();
        const actions = this._actionsMap.get(name);
        if (actions == null) {
            return;
        }
        for (const actionDef of actions) {
            actionDef.fn(...args);
        }
    }
    removeAction(name, priority, fn) {
        if (priority == null && fn == null) {
            this._actionsMap.delete(name);
            return;
        }
        const actions = this._actionsMap.get(name);
        if (actions == null) {
            return;
        }
        const updatedActions = actions.filter(def => !((priority === undefined || def.priority === priority) &&
            (fn === undefined || def.fn === fn)));
        this._actionsMap.set(name, updatedActions);
    }
    removeAllActions() {
        this._actionsMap.clear();
    }
}
const actionsManager = new ActionsManager();
function addAction(name, priority, fn) {
    actionsManager.addAction(name, priority, fn);
}
exports.addAction = addAction;
function doAction(name, ...args) {
    actionsManager.doAction(name, ...args);
}
exports.doAction = doAction;
function removeAction(name, priority, fn) {
    actionsManager.removeAction(name, priority, fn);
}
exports.removeAction = removeAction;
function removeAllActions() {
    actionsManager.removeAllActions();
}
exports.removeAllActions = removeAllActions;
function isTest() {
    // assume test suite is mocha, and check for it and describe
    return typeof globalThis.it === 'function' && typeof globalThis.describe === 'function';
}
exports.isTest = isTest;
const _initFunctions = [];
function onInit(fn) {
    /* istanbul ignore else */
    if (isTest()) {
        _initFunctions.push(fn);
    }
    else {
        fn();
    }
}
exports.onInit = onInit;
function doInit() {
    for (const fn of _initFunctions) {
        fn();
    }
}
exports.doInit = doInit;
const _shutdownFunctions = [];
function onShutdown(fn) {
    /* istanbul ignore else */
    if (isTest()) {
        _shutdownFunctions.push(fn);
    }
    else {
        // we don't need to run shutdown functions in C@E
    }
}
exports.onShutdown = onShutdown;
function doShutdown() {
    for (const fn of _shutdownFunctions) {
        fn();
    }
}
exports.doShutdown = doShutdown;
